<h1 style="color: red; text-align: center;">Assignment 8: Probability Concepts
(Questions)</h1>
<h2 style="color: red; text-align: center;">Instructor: Dr. Khaled Mohamad</h2>


# Q1: Feature Scaling Concepts
**Instructions:** Explain the following:
- Standardization
- Min-Max Scaling
- Robust scalling
- Why feature scaling is important
  
Write your answer below:

1. Standartization - is the way of scaling(also called z-score scaling), using the standard deviation and mean
   
    x=(x−mean)/std

After this mean would be equal to 0 and std to 1.

We can use it when we have a normal distibution.

2. Min-max scaling - is the way of scaling(also called z-score scaling), scales data to a fixed range [0, 1]

   x=(x−min)/(max−min)

Sensitive to outliers

3. Robust scaling - uses median and IQR (interquartile range)

    x=(x−median)/IQR

Works better with outliers.

Why feature scaling is important

Scaling ensures all features are on the same scale, preventing features with large values from dominating the model. It is especially important for distance-based and linear models like Logistic Regression.

# Q2: Load and Explore Dataset
**Instructions:**
- Load the breast cancer dataset from sklearn.datas
- Display shape and first 5 rows

In [12]:
from sklearn.datasets import load_breast_cancer
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler

data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

print("Shape:", X.shape)
X.head()

Shape: (569, 30)


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


# Q3: Train-Test Split
Split data into training and testing sets (80/20)

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Q4: Apply Feature Scaling
Apply:
- StandardScaler
- MinMaxScaler
- RobustScaler

In [17]:
scaler_std = StandardScaler()
scaler_mm = MinMaxScaler()
scaler_rb = RobustScaler()

In [37]:
X_train_std = scaler_std.fit_transform(X_train)
X_test_std = scaler_std.transform(X_test)

In [38]:
X_train_mm = scaler_mm.fit_transform(X_train)
X_test_mm = scaler_mm.transform(X_test)

In [39]:
X_train_rb = scaler_rb.fit_transform(X_train)
X_test_rb = scaler_rb.transform(X_test)

# Q5: Train Models
Train Logistic Regression on:
1. Unscaled data
2. Standardized scaled data
3. MinMax scaled data
4. Robust Scaled data

In [40]:
from sklearn.linear_model import LogisticRegression

In [41]:
model_raw = LogisticRegression(max_iter=10000)
model_raw.fit(X_train, y_train)

model_std = LogisticRegression(max_iter=10000)
model_std.fit(X_train_std, y_train)

model_mm = LogisticRegression(max_iter=10000)
model_mm.fit(X_train_mm, y_train)

model_rb = LogisticRegression(max_iter=10000)
model_rb.fit(X_train_rb, y_train)

LogisticRegression(max_iter=10000)

# Q6: Evaluate Each Model
Evaluate each model using the corresponding test set and print the following:
- Accuracy
- MAE
- RMSE

In [42]:
from sklearn.metrics import accuracy_score, mean_absolute_error, mean_squared_error

In [43]:
pred_raw = model_raw.predict(X_test)

acc_raw = accuracy_score(y_test, pred_raw)

mae_raw = mean_absolute_error(y_test, pred_raw)

rmse_raw = mean_squared_error(y_test, pred_raw)

print("Raw acc: ", acc_raw)
print("Raw mae: ", mae_raw)
print("Raw rmse: ", rmse_raw)

Raw acc:  0.956140350877193
Raw mae:  0.043859649122807015
Raw rmse:  0.043859649122807015


In [44]:
pred_std = model_std.predict(X_test_std)

acc_std = accuracy_score(y_test, pred_std)

mae_std = mean_absolute_error(y_test, pred_std)

rmse_std = mean_squared_error(y_test, pred_std)

print("Std acc: ", acc_std)
print("Std mae: ", mae_std)
print("Std rmse: ", rmse_std)

Std acc:  0.9736842105263158
Std mae:  0.02631578947368421
Std rmse:  0.02631578947368421


In [45]:
pred_mm = model_mm.predict(X_test_mm)

acc_mm = accuracy_score(y_test, pred_mm)

mae_mm = mean_absolute_error(y_test, pred_mm)

rmse_mm = mean_squared_error(y_test, pred_mm)

print("Mm acc: ", acc_mm)
print("Mm mae: ", mae_mm)
print("Mm rmse: ", rmse_mm)

Mm acc:  0.9824561403508771
Mm mae:  0.017543859649122806
Mm rmse:  0.017543859649122806


In [46]:
pred_rb = model_rb.predict(X_test_rb)

acc_rb = accuracy_score(y_test, pred_rb)

mae_rb = mean_absolute_error(y_test, pred_rb)

rmse_rb = mean_squared_error(y_test, pred_rb)

print("Rb acc: ", acc_rb)
print("Rb mae: ", mae_rb)
print("Rb rmse: ", rmse_rb)

Rb acc:  0.9824561403508771
Rb mae:  0.017543859649122806
Rb rmse:  0.017543859649122806


# Write down the results in a table (use markdown syntax for that). Use the following example as a guide

| Scaler Type     | Accuracy | MAE | RMSE |
|-----------------|----------|-----|------|
| No Scaling      | 0.95     | 0.04 | 0.04  |
| StandardScaler  | 0.97      | 0.03 | 0.03  |
| MinMaxScaler    | 0.98      | 0.02 | 0.02  |
| RobustScaler    | 0.98      | 0.02 | 0.02  |

# Q8: Reflection
Discuss:
- Impact of scaling on performance
- When scaling is necessary

## Impact of scaling on performance
Scaling improves model performance, especially for Logistic Regression, by ensuring features contribute equally. Models trained on scaled data usually converge faster and give better accuracy.

## When scaling is necessary
Scaling is necessary for:

- Logistic Regression
- SVM
- KNN

It is less important for tree-based models like Decision Trees or Random Forest.